### PR TITLE
P2p changes

### PR DESCRIPTION
**Pull request by Boog900**

### CODE REVIEW

**Comment by Boog900** [comment]
@hinto-janai This is ready, this PR is mainly just small changes I made to the current P2P code in #73 

I think the 2 biggest changes are, having a separate task to send pings to a peer and broadcasting is now done with a generic stream

File: p2p/monero-p2p/src/client/connector.rs [hunk 1]
```diff
@@ -1,39 +1,58 @@
+//! Connector
+//!
+//! This module handles connecting to peers and giving the sink/stream to the handshaker which will then
+//! perform a handshake and create a [`Client`].
+//!
+//! This is where outbound connections are crated.
```
Comment (hinto-janai) [inline p2p/monero-p2p/src/client/connector.rs:6]:
```suggestion
//! This is where outbound connections are created.
```

File: p2p/monero-p2p/src/client/connector.rs [hunk 2]
```diff
@@ -1,39 +1,58 @@
...
 use crate::{
     client::{Client, DoHandshakeRequest, HandShaker, HandshakeError, InternalPeerID},
-    AddressBook, ConnectionDirection, CoreSyncSvc, NetworkZone, PeerRequestHandler,
+    AddressBook, BroadcastMessage, ConnectionDirection, CoreSyncSvc, NetworkZone,
+    PeerRequestHandler, PeerSyncSvc,
 };
 
+/// A request to connect to a peer.
 pub struct ConnectRequest<Z: NetworkZone> {
+    /// The peers address
```
Comment (hinto-janai) [inline p2p/monero-p2p/src/client/connector.rs:26]: nit:
```suggestion
    /// The peer's address.
```

File: net/epee-encoding/src/container_as_blob.rs [hunk 3]
```diff
@@ -39,11 +39,7 @@ impl<T: Containerable + EpeeValue> EpeeValue for ContainerAsBlob<T> {
         }
 
         Ok(ContainerAsBlob(
-            bytes
-                .windows(T::SIZE)
-                .step_by(T::SIZE)
-                .map(T::from_bytes)
-                .collect(),
+            bytes.chunks(T::SIZE).map(T::from_bytes).collect(),
```
[the remark this answers is not in the document]
Comment (Boog900) [reply net/epee-encoding/src/container_as_blob.rs:42]: We check above this that the len is a multiple of `T::SIZE`

File: net/fixed-bytes/src/lib.rs [hunk 4]
```diff
@@ -101,6 +102,40 @@ impl<const N: usize> ByteArrayVec<N> {
...
+    ///
+    /// Afterwards self contains elements [0, at), and the returned Bytes contains elements [at, len).
+    ///
+    /// This is an O(1) operation that just increases the reference count and sets a few indices.
+    ///
+    /// # Panics
+    /// Panics if at > len.
+    pub fn split_off(&mut self, at: usize) -> Self {
+        Self(self.0.split_off(at * N))
+    }
```
[the remark this answers is not in the document]
Comment (Boog900) [reply net/fixed-bytes/src/lib.rs:116]: This is a byte array vec, i.e. a Vec<[u8; N]> with a backing representation of `Bytes` which is pretty much a Vec<u8> so all operations are scaled by N to account for this.

File: net/monero-wire/src/p2p/protocol.rs [hunk 5]
```diff
@@ -114,7 +114,7 @@ pub struct ChainResponse {
     /// Total Height
     pub total_height: u64,
     /// Cumulative Difficulty Low
-    pub cumulative_difficulty: u64,
+    pub cumulative_difficulty_low64: u64,
     /// Cumulative Difficulty High
     pub cumulative_difficulty_top64: u64,
```
[the remark this answers is not in the document]
Comment (Boog900) [reply net/monero-wire/src/p2p/protocol.rs:119]: no need IMO, I only did it here to match `cumulative_difficulty_top64`

File: net/monero-wire/src/p2p/protocol.rs [hunk 6]
```diff
@@ -125,11 +125,19 @@ pub struct ChainResponse {
     pub first_block: Bytes,
 }
 
+impl ChainResponse {
+    #[inline]
+    pub fn cumulative_difficulty(&self) -> u128 {
+        let cumulative_difficulty = self.cumulative_difficulty_top64 as u128;
+        cumulative_difficulty << 64 | self.cumulative_difficulty_low64 as u128
+    }
+}
```
[the remark this answers is not in the document]
Comment (Boog900) [reply net/monero-wire/src/p2p/protocol.rs:134]: This crate doesn't currently pull this crate in so I would rather keep this as is.

File: p2p/monero-p2p/src/client.rs [hunk 7]
```diff
@@ -86,25 +128,38 @@ impl<Z: NetworkZone> Service<PeerRequest> for Client<Z> {
...
+
+        self.permit = Some(permit);
+
+        Poll::Ready(Ok(()))
     }
 
     fn call(&mut self, request: PeerRequest) -> Self::Future {
+        let Some(permit) = self.permit.take() else {
+            panic!("poll_ready did not return ready before call to call")
+        };
```
[the remark this answers is not in the document]
Comment (Boog900) [reply p2p/monero-p2p/src/client.rs:152]: True, I changed this, not really a big issue though so I don't really want to go through the code to find these

File: p2p/address-book/src/peer_list.rs [hunk 8]
```diff
@@ -104,25 +104,21 @@ impl<Z: NetworkZone> PeerList<Z> {
                 })?;
                 let n = r.gen_range(0..addresses_with_block.len());
                 let peer = addresses_with_block[n];
-                if must_keep_peers.contains(&peer) {
-                    continue;
+                if !must_keep_peers.contains(&peer) {
+                    return self.remove_peer(&peer);
                 }
```
Comment (hinto-janai) [inline p2p/address-book/src/peer_list.rs:109]: Ah, I didn't edit https://github.com/Cuprate/cuprate/pull/118#discussion_r1585534711 in time I guess - I introduced a bug here, it should be reverted.

If this doesn't `continue` then the code below will execute, this is different behavior than before. Should be:
```rust
if must_keep_peers.contains(&peer) {
    continue;
}

return self.remove_peer(&peer);
```

Comment (Boog900) [reply p2p/address-book/src/peer_list.rs:109]: Ah, good spot!

**Review by hinto-janai — APPROVED** [review]

**Pull request merged**